### PR TITLE
Add Tsiveriotis Fernandes Lattice Engine

### PR DESCRIPTION
**Pull request by maittaeb**
Add the  Tsiveriotis Fernandes Lattice Engine for Convertible Bonds

Also added the Convertible Bonds example, ported from Quantlib

### CODE REVIEW

**Comment by igitur** [comment]
@amaggiulli has the final say, but I noticed some issues.

`ConvertibleBonds.csproj` has to be in the new .NET Core format. Look at https://github.com/amaggiulli/QLNet/blob/develop/src/Bonds/Bonds.csproj for an example. This is currently causing the CI build to fail.

Also, I suggest that you clean up the commit history.

**Comment by igitur** [comment]
And congratulations on your first pull request!

**Comment by amaggiulli** [comment]
Think error is for msbuild-sonarqube-runner version 2.3, I'll fix now and retry

**Comment by igitur** [comment]
@amaggiulli But still the new `.csproj` file should probably follow the current pattern for the example projects?

**Comment by amaggiulli** [comment]
@igitur correct , must be Net Core

File: src/QLNet/Instruments/Bonds/DiscretizedConvertible.cs [hunk 1]
```diff
@@ -0,0 +1,309 @@
...
+                if (arguments_.dividends[i].date() >= settlementDate)
+                {
+                    dividendValues_[i] = arguments_.dividends[i].amount() * process.riskFreeRate().currentLink().discount(arguments_.dividends[i].date());
+                }
+            }
+            DayCounter dayCounter = process.riskFreeRate().currentLink().dayCounter();
+            Date bondSettlement = arguments_.settlementDate;
+
+            stoppingTimes_ = new InitializedList<double>(arguments_.exercise.dates().Count(), 0.0);
+            for (var i = 0; i < stoppingTimes_.Count(); i++)
```
Comment (amaggiulli) [inline src/QLNet/Instruments/Bonds/DiscretizedConvertible.cs]: MAJOR Use 'Count' property here instead. [rule](https://sonarcloud.io/coding_rules#rule_key=csharpsquid%3AS2971)

File: src/QLNet/Instruments/Bonds/DiscretizedConvertible.cs [hunk 2]
```diff
@@ -0,0 +1,309 @@
...
+            }
+            for (var i = 0; i < callabilityTimes_.Count(); i++)
+            {
+                if (isOnTime(callabilityTimes_[i]))
+                {
+                    applyCallability(i, convertible);
+                }
+            }
+
+            for (var i = 0; i < couponTimes_.Count(); i++)
```
Comment (amaggiulli) [inline src/QLNet/Instruments/Bonds/DiscretizedConvertible.cs]: MAJOR Use 'Count' property here instead. [rule](https://sonarcloud.io/coding_rules#rule_key=csharpsquid%3AS2971)

File: src/QLNet/Instruments/Bonds/DiscretizedConvertible.cs [hunk 3]
```diff
@@ -0,0 +1,309 @@
...
+            DayCounter dayCounter = process.riskFreeRate().currentLink().dayCounter();
+            Date bondSettlement = arguments_.settlementDate;
+
+            stoppingTimes_ = new InitializedList<double>(arguments_.exercise.dates().Count(), 0.0);
+            for (var i = 0; i < stoppingTimes_.Count(); i++)
+            {
+                stoppingTimes_[i] = dayCounter.yearFraction(bondSettlement, arguments_.exercise.date(i));
+            }
+
+            callabilityTimes_ = new InitializedList<double>(arguments_.callabilityDates.Count(), 0.0);
```
Comment (amaggiulli) [inline src/QLNet/Instruments/Bonds/DiscretizedConvertible.cs]: MAJOR Use 'Count' property here instead. [rule](https://sonarcloud.io/coding_rules#rule_key=csharpsquid%3AS2971)

File: src/QLNet/Instruments/Bonds/DiscretizedConvertible.cs [hunk 4]
```diff
@@ -0,0 +1,309 @@
...
+        {
+            return dividendValues_;
+        }
+        public override void reset(int size)
+        {
+            // Set to bond redemption values
+            values_ = new Vector(size, Convert.ToDouble(arguments_.redemption));
+
+            //coupon amounts should be added when adjusting
+            // values_ = array(size,arguments_cashFlows.back()->amount());
```
Comment (amaggiulli) [inline src/QLNet/Instruments/Bonds/DiscretizedConvertible.cs]: MAJOR Remove this commented out code. [rule](https://sonarcloud.io/coding_rules#rule_key=csharpsquid%3AS125)

File: src/QLNet/Instruments/Bonds/DiscretizedConvertible.cs [hunk 5]
```diff
@@ -0,0 +1,309 @@
...
+            Date bondSettlement = arguments_.settlementDate;
+
+            stoppingTimes_ = new InitializedList<double>(arguments_.exercise.dates().Count(), 0.0);
+            for (var i = 0; i < stoppingTimes_.Count(); i++)
+            {
+                stoppingTimes_[i] = dayCounter.yearFraction(bondSettlement, arguments_.exercise.date(i));
+            }
+
+            callabilityTimes_ = new InitializedList<double>(arguments_.callabilityDates.Count(), 0.0);
+            for (var i = 0; i < callabilityTimes_.Count(); i++)
```
Comment (amaggiulli) [inline src/QLNet/Instruments/Bonds/DiscretizedConvertible.cs]: MAJOR Use 'Count' property here instead. [rule](https://sonarcloud.io/coding_rules#rule_key=csharpsquid%3AS2971)

File: src/QLNet/Instruments/Bonds/DiscretizedConvertible.cs [hunk 6]
```diff
@@ -0,0 +1,309 @@
...
+            {
+                callabilityTimes_[i] = dayCounter.yearFraction(bondSettlement, arguments_.callabilityDates[i]);
+            }
+
+            couponTimes_ = new InitializedList<double>(arguments_.couponDates.Count(), 0.0);
```
Comment (amaggiulli) [inline src/QLNet/Instruments/Bonds/DiscretizedConvertible.cs]: MAJOR Use 'Count' property here instead. [rule](https://sonarcloud.io/coding_rules#rule_key=csharpsquid%3AS2971)

File: src/QLNet/Instruments/Bonds/DiscretizedConvertible.cs [hunk 7]
```diff
@@ -0,0 +1,309 @@
...
+            for (var i = 0; i < couponTimes_.Count(); i++)
```
Comment (amaggiulli) [inline src/QLNet/Instruments/Bonds/DiscretizedConvertible.cs]: MAJOR Use 'Count' property here instead. [rule](https://sonarcloud.io/coding_rules#rule_key=csharpsquid%3AS2971)

File: src/QLNet/Instruments/Bonds/DiscretizedConvertible.cs [hunk 8]
```diff
@@ -0,0 +1,309 @@
...
+            {
+                couponTimes_[i] = dayCounter.yearFraction(bondSettlement, arguments_.couponDates[i]);
+            }
+
+            dividendTimes_ = new InitializedList<double>(arguments_.dividendDates.Count(), 0.0);
```
Comment (amaggiulli) [inline src/QLNet/Instruments/Bonds/DiscretizedConvertible.cs]: MAJOR Use 'Count' property here instead. [rule](https://sonarcloud.io/coding_rules#rule_key=csharpsquid%3AS2971)

File: src/QLNet/Instruments/Bonds/DiscretizedConvertible.cs [hunk 9]
```diff
@@ -0,0 +1,309 @@
...
+            for (var i = 0; i < dividendTimes_.Count(); i++)
```
Comment (amaggiulli) [inline src/QLNet/Instruments/Bonds/DiscretizedConvertible.cs]: MAJOR Use 'Count' property here instead. [rule](https://sonarcloud.io/coding_rules#rule_key=csharpsquid%3AS2971)

File: src/QLNet/Instruments/Bonds/DiscretizedConvertible.cs [hunk 10]
```diff
@@ -0,0 +1,309 @@
...
+            {
+                dividendTimes_[i] = dayCounter.yearFraction(bondSettlement, arguments_.dividendDates[i]);
+            }
+
+            if (!grid.empty())
+            {
+                // adjust times to grid 
+                for (var i = 0; i < stoppingTimes_.Count(); i++)
```
Comment (amaggiulli) [inline src/QLNet/Instruments/Bonds/DiscretizedConvertible.cs]: MAJOR Use 'Count' property here instead. [rule](https://sonarcloud.io/coding_rules#rule_key=csharpsquid%3AS2971)

File: src/QLNet/Instruments/Bonds/DiscretizedConvertible.cs [hunk 11]
```diff
@@ -0,0 +1,309 @@
...
+
+
+            // Claculate blended discount rate to be used on roll back .
+            for (var j = 0; j < values_.Count(); j++)
+            {
+                spreadAdjustedRate_[j] = conversionProbability_[j] * riskFreeRate.value() + (1 - conversionProbability_[j]) * (riskFreeRate.value() + creditSpread.currentLink().value());
+            }
+        }
+
+        protected override void postAdjustValuesImpl()
```
Comment (amaggiulli) [inline src/QLNet/Instruments/Bonds/DiscretizedConvertible.cs]: CRITICAL Refactor this method to reduce its Cognitive Complexity from 18 to the 15 allowed. [rule](https://sonarcloud.io/coding_rules#rule_key=csharpsquid%3AS3776)

File: src/QLNet/Instruments/Bonds/DiscretizedConvertible.cs [hunk 12]
```diff
@@ -0,0 +1,309 @@
...
+using System.Linq;
+
+
+
+namespace QLNet
+{
+    
+    public class DiscretizedConvertible : DiscretizedAsset
+    {
+        private ConvertibleBond.option.arguments arguments_;
```
Comment (amaggiulli) [inline src/QLNet/Instruments/Bonds/DiscretizedConvertible.cs]: MAJOR Make 'arguments_' 'readonly'. [rule](https://sonarcloud.io/coding_rules#rule_key=csharpsquid%3AS2933)

File: src/QLNet/Instruments/Bonds/DiscretizedConvertible.cs [hunk 13]
```diff
@@ -0,0 +1,309 @@
...
+                double payoff = Convert.ToDouble(arguments_.conversionRatio) * grid[j];
+                if (values_[j] <= payoff)
+                {
+                    values_[j] = payoff;
+                    conversionProbability_[j] = 1.0;
+                }
+            }
+        }
+
+        public void applyCallability(int i, bool convertible)
```
Comment (amaggiulli) [inline src/QLNet/Instruments/Bonds/DiscretizedConvertible.cs]: CRITICAL Refactor this method to reduce its Cognitive Complexity from 20 to the 15 allowed. [rule](https://sonarcloud.io/coding_rules#rule_key=csharpsquid%3AS3776)

File: src/QLNet/Instruments/Bonds/DiscretizedConvertible.cs [hunk 14]
```diff
@@ -0,0 +1,309 @@
...
+
+
+
+namespace QLNet
+{
+    
+    public class DiscretizedConvertible : DiscretizedAsset
+    {
+        private ConvertibleBond.option.arguments arguments_;
+        private GeneralizedBlackScholesProcess process_;
```
Comment (amaggiulli) [inline src/QLNet/Instruments/Bonds/DiscretizedConvertible.cs]: MAJOR Make 'process_' 'readonly'. [rule](https://sonarcloud.io/coding_rules#rule_key=csharpsquid%3AS2933)

File: src/QLNet/Instruments/Bonds/DiscretizedConvertible.cs [hunk 15]
```diff
@@ -0,0 +1,309 @@
...
+
+            if (!grid.empty())
+            {
+                // adjust times to grid 
+                for (var i = 0; i < stoppingTimes_.Count(); i++)
+                {
+                    stoppingTimes_[i] = grid.closestTime(stoppingTimes_[i]);
+                }
+
+                for (var i = 0; i < couponTimes_.Count(); i++)
```
Comment (amaggiulli) [inline src/QLNet/Instruments/Bonds/DiscretizedConvertible.cs]: MAJOR Use 'Count' property here instead. [rule](https://sonarcloud.io/coding_rules#rule_key=csharpsquid%3AS2971)

File: src/QLNet/Instruments/Bonds/DiscretizedConvertible.cs [hunk 16]
```diff
@@ -0,0 +1,309 @@
...
+
+
+namespace QLNet
+{
+    
+    public class DiscretizedConvertible : DiscretizedAsset
+    {
+        private ConvertibleBond.option.arguments arguments_;
+        private GeneralizedBlackScholesProcess process_;
+        private List<double> stoppingTimes_;
```
Comment (amaggiulli) [inline src/QLNet/Instruments/Bonds/DiscretizedConvertible.cs]: MAJOR Make 'stoppingTimes_' 'readonly'. [rule](https://sonarcloud.io/coding_rules#rule_key=csharpsquid%3AS2933)

File: src/QLNet/Instruments/Bonds/DiscretizedConvertible.cs [hunk 17]
```diff
@@ -0,0 +1,309 @@
...
+        private List<double> callabilityTimes_;
```
Comment (amaggiulli) [inline src/QLNet/Instruments/Bonds/DiscretizedConvertible.cs]: MAJOR Make 'callabilityTimes_' 'readonly'. [rule](https://sonarcloud.io/coding_rules#rule_key=csharpsquid%3AS2933)

File: src/QLNet/Instruments/Bonds/DiscretizedConvertible.cs [hunk 18]
```diff
@@ -0,0 +1,309 @@
...
+        private List<double> couponTimes_;
```
Comment (amaggiulli) [inline src/QLNet/Instruments/Bonds/DiscretizedConvertible.cs]: MAJOR Make 'couponTimes_' 'readonly'. [rule](https://sonarcloud.io/coding_rules#rule_key=csharpsquid%3AS2933)

File: src/QLNet/Instruments/Bonds/DiscretizedConvertible.cs [hunk 19]
```diff
@@ -0,0 +1,309 @@
...
+        private List<double> dividendTimes_;
```
Comment (amaggiulli) [inline src/QLNet/Instruments/Bonds/DiscretizedConvertible.cs]: MAJOR Make 'dividendTimes_' 'readonly'. [rule](https://sonarcloud.io/coding_rules#rule_key=csharpsquid%3AS2933)

File: src/QLNet/Instruments/Bonds/DiscretizedConvertible.cs [hunk 20]
```diff
@@ -0,0 +1,309 @@
...
+                {
+                    stoppingTimes_[i] = grid.closestTime(stoppingTimes_[i]);
+                }
+
+                for (var i = 0; i < couponTimes_.Count(); i++)
+                {
+                    couponTimes_[i] = grid.closestTime(couponTimes_[i]);
+                }
+
+                for (var i = 0; i < dividendTimes_.Count(); i++)
```
Comment (amaggiulli) [inline src/QLNet/Instruments/Bonds/DiscretizedConvertible.cs]: MAJOR Use 'Count' property here instead. [rule](https://sonarcloud.io/coding_rules#rule_key=csharpsquid%3AS2971)

File: src/QLNet/Instruments/Bonds/DiscretizedConvertible.cs [hunk 21]
```diff
@@ -0,0 +1,309 @@
...
+    public class DiscretizedConvertible : DiscretizedAsset
+    {
+        private ConvertibleBond.option.arguments arguments_;
+        private GeneralizedBlackScholesProcess process_;
+        private List<double> stoppingTimes_;
+        private List<double> callabilityTimes_;
+        private List<double> couponTimes_;
+        private List<double> dividendTimes_;
+
+        public Vector conversionProbability_;
```
Comment (amaggiulli) [inline src/QLNet/Instruments/Bonds/DiscretizedConvertible.cs]: MINOR Make this field 'private' and encapsulate it in a 'public' property. [rule](https://sonarcloud.io/coding_rules#rule_key=csharpsquid%3AS1104)

File: src/QLNet/Instruments/Bonds/DiscretizedConvertible.cs [hunk 22]
```diff
@@ -0,0 +1,309 @@
...
+        public Vector spreadAdjustedRate_;
```
Comment (amaggiulli) [inline src/QLNet/Instruments/Bonds/DiscretizedConvertible.cs]: MINOR Make this field 'private' and encapsulate it in a 'public' property. [rule](https://sonarcloud.io/coding_rules#rule_key=csharpsquid%3AS1104)

File: src/QLNet/Instruments/Bonds/DiscretizedConvertible.cs [hunk 23]
```diff
@@ -0,0 +1,309 @@
...
+        public Vector dividendValues_;
```
Comment (amaggiulli) [inline src/QLNet/Instruments/Bonds/DiscretizedConvertible.cs]: MINOR Make this field 'private' and encapsulate it in a 'public' property. [rule](https://sonarcloud.io/coding_rules#rule_key=csharpsquid%3AS1104)

File: src/QLNet/Instruments/Bonds/DiscretizedConvertible.cs [hunk 24]
```diff
@@ -0,0 +1,309 @@
...
+                {
+                    couponTimes_[i] = grid.closestTime(couponTimes_[i]);
+                }
+
+                for (var i = 0; i < dividendTimes_.Count(); i++)
+                {
+                    dividendTimes_[i] = grid.closestTime(dividendTimes_[i]);
+                }
+
+                for (var i = 0; i < callabilityTimes_.Count(); i++)
```
Comment (amaggiulli) [inline src/QLNet/Instruments/Bonds/DiscretizedConvertible.cs]: MAJOR Use 'Count' property here instead. [rule](https://sonarcloud.io/coding_rules#rule_key=csharpsquid%3AS2971)

File: src/QLNet/Instruments/Bonds/DiscretizedConvertible.cs [hunk 25]
```diff
@@ -0,0 +1,309 @@
...
+
+        public Vector conversionProbability_;
+        public Vector spreadAdjustedRate_;
+        public Vector dividendValues_;
+
+
+        public ConvertibleBond.option.arguments arguments() { return arguments_; }
+        public GeneralizedBlackScholesProcess process() { return process_; }
+
+        public DiscretizedConvertible(ConvertibleBond.option.arguments args, GeneralizedBlackScholesProcess process, TimeGrid grid)
```
Comment (amaggiulli) [inline src/QLNet/Instruments/Bonds/DiscretizedConvertible.cs]: CRITICAL Refactor this constructor to reduce its Cognitive Complexity from 16 to the 15 allowed. [rule](https://sonarcloud.io/coding_rules#rule_key=csharpsquid%3AS3776)

File: src/QLNet/Instruments/Bonds/DiscretizedConvertible.cs [hunk 26]
```diff
@@ -0,0 +1,309 @@
...
+                    }
+                    break;
+                case Exercise.Type.European:
+                    if (isOnTime(stoppingTimes_[0]))
+                    {
+                        convertible = true;
+                    }
+                    break;
+                case Exercise.Type.Bermudan:
+                    for (var i = 0; i < stoppingTimes_.Count(); i++)
```
Comment (amaggiulli) [inline src/QLNet/Instruments/Bonds/DiscretizedConvertible.cs]: MAJOR Use 'Count' property here instead. [rule](https://sonarcloud.io/coding_rules#rule_key=csharpsquid%3AS2971)

File: src/QLNet/Instruments/Bonds/DiscretizedConvertible.cs [hunk 27]
```diff
@@ -0,0 +1,309 @@
...
+                        }
+                    }
+                    break;
+                default:
+                    {
+                        Utils.QL_FAIL("invalid option type ");
+                    }
+                    break;
+            }
+            for (var i = 0; i < callabilityTimes_.Count(); i++)
```
Comment (amaggiulli) [inline src/QLNet/Instruments/Bonds/DiscretizedConvertible.cs]: MAJOR Use 'Count' property here instead. [rule](https://sonarcloud.io/coding_rules#rule_key=csharpsquid%3AS2971)

File: src/QLNet/Instruments/Bonds/DiscretizedConvertible.cs [hunk 28]
```diff
@@ -0,0 +1,309 @@
...
+            {
+                if (arguments_.dividends[i].date() >= settlementDate)
+                {
+                    dividendValues_[i] = arguments_.dividends[i].amount() * process.riskFreeRate().currentLink().discount(arguments_.dividends[i].date());
+                }
+            }
+            DayCounter dayCounter = process.riskFreeRate().currentLink().dayCounter();
+            Date bondSettlement = arguments_.settlementDate;
+
+            stoppingTimes_ = new InitializedList<double>(arguments_.exercise.dates().Count(), 0.0);
```
Comment (amaggiulli) [inline src/QLNet/Instruments/Bonds/DiscretizedConvertible.cs]: MAJOR Use 'Count' property here instead. [rule](https://sonarcloud.io/coding_rules#rule_key=csharpsquid%3AS2971)

**Comment by amaggiulli** [comment]
SonarQube analysis reported 32 issues
* ![CRITICAL][CRITICAL] 3 critical
* ![MAJOR][MAJOR] 26 major
* ![MINOR][MINOR] 3 minor

Watch the comments in this conversation to review them.

#### 4 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [bsmlattice.cs#L26](https://github.com/amaggiulli/QLNet/blob/d13568a6e23888c4d2d11b0a0470fce5c8a1ab45/src/QLNet/Methods/lattices/bsmlattice.cs#L26): 'T' is not used in the class. [rule](https://sonarcloud.io/coding_rules#rule_key=csharpsquid%3AS2326)
1. ![MAJOR][MAJOR] [bsmlattice.cs#L110](https://github.com/amaggiulli/QLNet/blob/d13568a6e23888c4d2d11b0a0470fce5c8a1ab45/src/QLNet/Methods/lattices/bsmlattice.cs#L110): Make 'discount_' 'readonly'. [rule](https://sonarcloud.io/coding_rules#rule_key=csharpsquid%3AS2933)
1. ![MAJOR][MAJOR] [bsmlattice.cs#L111](https://github.com/amaggiulli/QLNet/blob/d13568a6e23888c4d2d11b0a0470fce5c8a1ab45/src/QLNet/Methods/lattices/bsmlattice.cs#L111): Make 'pd_' 'readonly'. [rule](https://sonarcloud.io/coding_rules#rule_key=csharpsquid%3AS2933)
1. ![MAJOR][MAJOR] [bsmlattice.cs#L111](https://github.com/amaggiulli/QLNet/blob/d13568a6e23888c4d2d11b0a0470fce5c8a1ab45/src/QLNet/Methods/lattices/bsmlattice.cs#L111): Make 'pu_' 'readonly'. [rule](https://sonarcloud.io/coding_rules#rule_key=csharpsquid%3AS2933)

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by amaggiulli** [comment]
AppVeyor build fixed and is good , @maittaeb some work to do anyway , you can ignore 'readonly' issues but the others have to be fixed. Example project must be done in the new csproj format and remember to run format_code.bat before merging. Finally squash your pull request to not have all synch commits , if you need help on this you can ask @igitur to help you . Have fun !

**Comment by maittaeb** [comment]
Okay thank you, I'll try to figure out the issues you mentioned.

**Pull request closed without merging**